src/sentry/services/hybrid_cloud/rpc/serialization.py

```diff
@@ -0,0 +1,336 @@
+
+
+class DataClassSerializer(JsonSerializer):
+    fields: Mapping[str, JsonSerializer]
+    constructor: Any
+
+    def __init__(self, fields: Mapping[str, JsonSerializer], constructor: Any):
+        self.fields = fields
+        self.constructor = constructor
+
+    def schema_type(self) -> Mapping[str, Any]:
+        return dict(type="object", properties={k: v.schema_type() for k, v in self.fields.items()})
+
+    def to_json(self, value: Any) -> Any:
+        result: Any = getattr(value, "__extra_fields__", None) or {}
+        for k, s in self.fields.items():
+            v = getattr(value, k)
+            if v is Unset:
+                continue
+            result[k] = s.to_json(v)
+        return result
+
+    def from_json(self, value: Any) -> Any:
+        kwds: MutableMapping[str, Any] = {}
+        extra_fields: MutableMapping[str, Any] = {}
+        for k, v in value.items():
+            if k in self.fields:
+                kwds[k] = self.fields[k].from_json(v)
+            else:
+                extra_fields[k] = v
+        result = self.constructor(**kwds)
+        setattr(result, "__extra_fields__", extra_fields)
+        return result
+
+
+@dataclasses.dataclass
+class Entry:
+    k: str
+    v: object
+
+
+class EntrySerializer(DataClassSerializer):
+    inner: JsonSerializer
+
+    def __init__(self, inner: JsonSerializer):
+        self.inner = inner
+        super().__init__({"k": str_serializer, "v": inner}, Entry)
+
+    def schema_type(self) -> Mapping[str, Any]:
+        return dict(
+            type="object",
+            properties=dict(
+                k=str_serializer.schema_type(),
+                v=self.inner.schema_type(),
+            ),
+            required=["k", "v"],
+        )
+
+
+class NullableSerializer(JsonSerializer):
+    inner: JsonSerializer
+
+    def __init__(self, inner: JsonSerializer):
+        self.inner = inner
+
+    def schema_type(self) -> Mapping[str, Any]:
+        return dict(nullable=True, **self.inner.schema_type())
+
+    def to_json(self, value: Any) -> Any:
+        if value is None:
+            return None
+        return self.inner.to_json(value)
+
+    def from_json(self, value: Any) -> Any:
+        if value is None:
+            return None
+        return self.inner.from_json(value)
+
+
+class MapSerializer(ListSerializer):
+    def __init__(self, inner: JsonSerializer):
+        super().__init__(lambda values: {e.k: e.v for e in values}, EntrySerializer(inner))
+
+    def to_json(self, value: Any) -> Any:
+        return super().to_json(Entry(k=k, v=v) for k, v in value.items())
+
+
+_id = lambda i: i
+
+int_serializer = ScalarSerializer(dict(type="integer", format="int64"), _id, _id)
+str_serializer = ScalarSerializer(dict(type="string"), _id, _id)
+float_serializer = ScalarSerializer(dict(type="number"), _id, _id)
+bool_serializer = ScalarSerializer(dict(type="boolean"), _id, _id)
+opaque_serializer = ScalarSerializer(dict(type="object"), _id, _id)
```

Comment:
Should we be doing casting for scalars when parsing data from json payloads? While we'll have typechecks on both sides we shouldn't trust the data coming off the wire.

Reply:
Valid point.  I honestly, hope we can borrow a library to do this, and I think pydantic might be a good choice.  This "quick hack" was demonstrative of needs and not intended to be a final implementation.